android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesFactory.kt

```diff
@@ -124,6 +126,54 @@ class RulesFactory @Inject constructor(val configurationRegistry: ConfigurationR
         Locale.getDefault(),
         "{{${value.translationPropertyKey()}}}"
       )
+
+    /**
+     * This method retrieves a list of relatedResources for a given resource from the facts map It
+     * fetches a list of facts of the given [relatedResourceType] then iterates through this list in
+     * order to return a list of all resources whose subject reference matches the logical Id of the
+     * [resource]
+     *
+     * [resource]
+     * - The parent resource for which the related resources will be retrieved [relatedResourceType]
+     * - The ResourceType the relatedResources belong to [fhirPathExpression]
+     * - A fhir path expression used to retrieve the subject reference Id from the related resources
+     */
+    fun retrieveRelatedResources(
+      resource: Resource,
+      relatedResourceType: String,
+      fhirPathExpression: String
+    ): List<Resource> {
+      val value = facts.getFact(relatedResourceType).value as ArrayList<Resource>
+
+      return value.filter {
+        resource.logicalId ==
+          fhirPathDataExtractor
+            .extractValue(it, fhirPathExpression)
+            .substringAfterLast(delimiter = '/', missingDelimiterValue = "")
+      }
+    }
+
+    /**
+     * This method retrieve a parentResources for a given relatedResource from the facts map It
+     * fetches a list of facts of the given [parentResourceType] then iterates through this list in
+     * order to return a resource whose logical id matches the given [parentResourceId]
+     *
+     * [parentResourceId]
+     * - The logical Id of the parentResource [parentResourceType]
+     * - The ResourceType the parentResources belong to [fhirPathExpression]
+     * - A fhir path expression used to retrieve the logical Id from the parent resources
+     */
+    fun retrieveParentResource(
+      parentResourceId: String,
+      parentResourceType: String,
+      fhirPathExpression: String
+    ): Resource? {
```

Comment:
Update this function's signature to accept, child `resource`, `parentResouceType` and `fhirPathExpression` for extracting the needed ID from the child resource.